### ISSUE TITLE
Cartesian coordinates of locations

### ISSUE THREAD

**Issue by cdondrup** [LAAS-HRI/semantic_route_description#3]
In order to figure out if a person has to go left right or straight at a certain point along the route, I would need to be able to get the cartesian coordinates of that location. Is there a way to get those using ontologenius or could they be included in the route information?

**Comment by sarthou** [comment]
Cartesian coordinates are indeed the easiest way to get accurate information to determine whether a person should go to the right or left at a certain point of the route. However, I think this type of information will be difficult to describe for large areas.
If you do not find other solutions, it is always possible to describe the coordinates in the ontology via the data properties.

**Comment by sarthou** [comment]
To create data properties, you can view [here](https://github.com/sarthou/ontologenius/wiki/Create-ontology#data-property).
To get the data once they are in ontologenius, it will be [here](https://github.com/sarthou/ontologenius/wiki/Individuals-exploration). In this part, when you query a property, it queries the properties of data and objects.

**Comment by sarthou** [comment]
Here is an example of a data property [definition](https://github.com/sarthou/ontologenius/blob/9900c48290d3b8595e641f7ab3a1a646836c5932/files/positionProperty.owl#L212) and [attribution](https://github.com/sarthou/ontologenius/blob/9900c48290d3b8595e641f7ab3a1a646836c5932/files/testIndividuals.owl#L85) to an individual

**Comment by cdondrup** [comment]
Follow-up question, I understand the problem of using coordinates for large areas and lots of shops so if I would like to avoid that, is there any way of finding if something is to the left, right, or straight in an `openspace`? 

As far as I understand, `corridor`s have a clearly defined left and right whereas `opensapce`s do not. So looking at this for example:

![os_hall](https://user-images.githubusercontent.com/4293484/44407364-d3433380-a555-11e8-81be-ae1039337939.png)

Just shows an arbitrary spacial arrangement of an ordered list of places. From what I understand from [this function](https://github.com/LAAS-HRI/route_verbalization/blob/master/src/mapManipulators/MapReader.cpp#L120-L173), you only get an ordered list of all the shops based on the `isAtRightOf` property but cannot tell anything about the spatial arrangement. The image above, however, would imply that the `elevator_1` is opposite of `door_h20` but that is just the way it's drawn and there is no way of knowing that from the current ontology, right? Therefore, I cannot produce sentences like: "Go through the door and then straight ahead towards the elevator."

**Comment by sarthou** [comment]
For openspaces, on some shop, you will find an `isInFrontOf` or` hasInFront` property. It is thanks to this property that you can know if a shop is on the right, left or front.

On your picture, if you're coming from door_h20 and want to go up the stairs, the door_h20 is in front of gf_toilets so you can deduce that the stairs are on your left.

**Comment by sarthou** [comment]
That's right, the solution is the `isInFrontOf` property. If you have the latest version of the package semantic_route_description, you should have a door in front of intersection 34.

**Comment by cdondrup** [comment]
But I can see that this is not the case for every corridor. So not every intersection has something in front of it. Also, if you have a layout like this:

```
                                   B1
                                   |
                                   |
                                   |
                                   |
B2----------------------------------------------------------------------E2
                                   |
                                   |
                                   |
                                   |
                                   |
                                   E1
```
you would have intersection c12 on either side of both corridors, right? Or would you split the corridors so it becomes:

```
                                   B1
                                   |
                                   |
                                   |
                                   E1
B2-------------------------------E2 B4-------------------------------E4
                                   B3
                                   |
                                   |
                                   |
                                   |
                                   E3
```
?

My concern is just that it has to be consistent across all the corridors and spaces. Otherwise, the whole reasoning over the ontology turns into guesswork.

Coming back to the example I gave earlier with burger king, in this case, I can just look up `ff_c34_intersection:isInFrontOf` and see if burger king is either connected to the intersection or connected to whatever is infront of the intersection and then see if it is left or right of it. But if there is nothing in front of the intersection, that logic will already fail. Obviously, I can just try and if it fails I just don't have the information and have to compensate for it but I don't want to have to try 5 different approaches with increasing amount of guesswork of getting the same information. Ideally, I would like to check 

1. if the target is at the beginning/end of the corridor 
1. if it is left or right of the interface
1. if it is left or right of what is opposite of the interface

This might fail at multiple points, e.g. if there are gaps between any of the places, or if there is nothing opposite of the intersection. This could be prevented by introducing empty spaces that just connect things so we have a continuous chain of places along each side of a corridor and by always having something opposite of the intersection either a shop if there is one or a dummy place like the empty ones.

**Comment by cdondrup** [comment]
Thanks, that would be awesome!

**Comment by cdondrup** [comment]
Hi, any progress on this?

**Comment by cdondrup** [comment]
Great, thank you! Let me know when it's done. Have to start testing things.

**Comment by sarthou** [comment]
Here the [first version](https://github.com/LAAS-HRI/semantic_route_description/commit/efd780190cfc2b4d72ced3d01f0b15a97116027b).
If you find some problems, do not hesitate to fork it and make pull-requests

**Comment by cdondrup** [comment]
I just had a look at the images drawn by the rout_verbalization and can see that there are still unconnected shops, etc. Are you planning to fill those gaps with empty spaces as we discussed? Otherwise, I won't be able to reliably generate route descriptions.

**Comment by sarthou** [comment]
I put empty spaces in the description. When there is no connection on the images, it's just that my drawing software is a small tool that is not smart enough to estimate the size of each element and therefore fill the holes. However, if you see places where you would need empty space, let me know, I'll put them.
